data_parallel_tensor.py

```diff
@@ -0,0 +1,396 @@
+from enum import Enum, auto
+from typing import Any, List, Optional, Tuple, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch._C import NoneType, device
+from torch._utils import _get_all_device_indices
+from torch.cuda import comm
+from torch.utils._pytree import tree_map
+
+torch.__future__.set_overwrite_module_params_on_conversion(True)
+import concurrent.futures as futures
+
+_ = torch.manual_seed(0)
+aten = torch.ops.aten
+NUM_DEVICES = 8
+PARALLEL_DISPATCH = False
+ALL_REDUCE = True
+
+
+class DPTensorType(Enum):
+    replicated = auto()  # This tensor will be replicated across all the devices
+    distributed_batch = (
+        auto()
+    )  # This tensor will be sharded along the first/batch dimension across
+    # the devices, NOTE: only equal chunk sizes are supported
+    distributed = (
+        auto()
+    )  # This is a list of tensors, each of which rests on different devices
+
+
+class DataParallelTensor(torch.Tensor):
+    # This class is a tensor subclass that stores a list of tensors with the aim
+    # DataParallelTensors(DPT) are categorized in three ways
+    # 1) replicated: When a single tensor is supplied, it is replicated across
+    #   all the devices by using broadcast
+    # 2) distributed: DPT can also be initialized by supplying a list/tuple of tensors
+    #   if the elements rest on different devices, they will just be wrapped in DPT
+    #   else the elements are scattered to different devices
+    # 3) distributed batch: This type of DPT tensor is created by sharding the input tensor across
+    #   a specified sharding dimension (default: 0). Currently only equal chunk sizes are supported.
+
+    elem: List[torch.Tensor]
+
+    if torch.cuda.is_available():
+        # device_ids: List[int] = _get_all_device_indices()
+        device_ids = [i for i in range(NUM_DEVICES)]
+        if PARALLEL_DISPATCH:
+            num_threads: int = len(device_ids)
+            threadpool: futures.ThreadPoolExecutor = futures.ThreadPoolExecutor(
+                max_workers=num_threads
+            )
+    __slots__ = ["elem"]
+
+    @staticmethod
+    def __new__(
+        cls,
+        elem: Union[torch.Tensor, List[torch.Tensor], Tuple[torch.Tensor]],
+        func: Optional[Any] = None,
+        dpt_type: DPTensorType = DPTensorType.replicated,
+        batch_dim: Optional[int] = 0,
+    ):
+
+        if dpt_type == DPTensorType.replicated:
+            # NOTE: If the input is None, we return None
+            if elem is None:
+                return None
+            assert isinstance(elem, torch.Tensor)
+            # NOTE: For handling meta tensors, if the device of an input tensor is meta,
+            # we just return the first element in such a list/tuple
+            if elem.device == device("meta"):
+                return elem
+
+            with torch.no_grad():
```

Comment:
Just for curiosity. What happens if we don't have `no_grad`?

Reply:
If the tensor being replicated has it's required grad field set then the broadcast/scatter operation will be recorded in its graph and will cause the gradients to flow to the base tenor when calling backwards. Also it cannot be released from memory since it's reference is retained now.